x/ecocredit/server/basket/put.go

```diff
@@ -2,11 +2,181 @@ package basket
+			minStartDate = blockTime.Add(window)
+		}
+
+		if batchInfo.StartDate.Before(minStartDate) {
+			return sdkerrors.ErrInvalidRequest.Wrapf("cannot put a credit from a batch with start time %s "+
+				"into a basket that requires a min start time of %s", batchInfo.StartDate.String(), minStartDate.String())
+		}
+
+	}
+
+	// check credit class match
+	found, err := k.stateStore.BasketClassStore().Has(ctx, basket.Id, batchInfo.ClassId)
+	if err != nil {
+		return err
+	}
+	if !found {
+		return sdkerrors.ErrInvalidRequest.Wrapf("credit class %s is not allowed in this basket", batchInfo.ClassId)
+	}
+
+	// check credit type match
+	requiredCreditType := basket.CreditTypeName
+	res2, err := k.ecocreditKeeper.ClassInfo(ctx, &ecocredit.QueryClassInfoRequest{ClassId: batchInfo.ClassId})
+	if err != nil {
+		return err
+	}
+	gotCreditType := res2.Info.CreditType.Name
+	if requiredCreditType != gotCreditType {
+		return sdkerrors.ErrInvalidRequest.Wrapf("cannot use credit of type %s in a basket that requires credit type %s", gotCreditType, requiredCreditType)
+	}
+	return nil
+}
+
+// updateBalances updates the balance of the user in the legacy KVStore as well as the basket's balance in the ORM.
+func (k Keeper) updateBalances(ctx context.Context, sender sdk.AccAddress, amt regenmath.Dec, basket *basketv1.Basket, batchInfo *ecocredit.BatchInfo) error {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	store := sdkCtx.KVStore(k.storeKey)
+
+	// update the user balance
+	userBalanceKey := ecocredit.TradableBalanceKey(sender, ecocredit.BatchDenomT(batchInfo.BatchDenom))
+	userBalance, err := ecocredit.GetDecimal(store, userBalanceKey)
+	if err != nil {
+		return err
+	}
+	newUserBalance, err := regenmath.SafeSubBalance(userBalance, amt)
+	if err != nil {
+		return err
+	}
+	ecocredit.SetDecimal(store, userBalanceKey, newUserBalance)
+
+	// update basket balance with amount sent
+	var bal *basketv1.BasketBalance
+	bal, err = k.stateStore.BasketBalanceStore().Get(ctx, basket.Id, batchInfo.BatchDenom)
+	if err != nil {
+		if ormerrors.IsNotFound(err) {
+			bal = &basketv1.BasketBalance{
+				BasketId:       basket.Id,
+				BatchDenom:     batchInfo.BatchDenom,
+				Balance:        amt.String(),
+				BatchStartDate: timestamppb.New(*batchInfo.StartDate),
+			}
+		} else {
+			return err
+		}
+	} else {
+		newBalance, err := regenmath.NewPositiveFixedDecFromString(bal.Balance, basket.Exponent)
+		if err != nil {
+			return err
+		}
+		newBalance, err = newBalance.Add(amt)
+		if err != nil {
+			return err
+		}
+		bal.Balance = newBalance.String()
+	}
+	if err = k.stateStore.BasketBalanceStore().Save(ctx, bal); err != nil {
+		return err
+	}
+	return nil
+}
+
+// calculateTokens calculates the tokens to award to the depositor
+func calculateTokens(creditAmt regenmath.Dec, exp uint32, denom string) (sdk.Coins, error) {
+	var coins sdk.Coins
+	multiplier := regenmath.NewDecFinite(10, int32(exp))
```

Comment:
From my PR pretty sure this should be:
```suggestion
	multiplier := regenmath.NewDecFinite(1, int32(exp))
```

Reply:
yup, fixed